File: frontend/src/types/vton.ts
import { MannequinParams } from "./body";
import { FitZone } from "./product";

export type VtonRunMode = "mock" | "external" | "auto";
export type VtonTaskState = "queued" | "running" | "succeeded" | "failed";
export type VtonRenderMethod = "NEURAL_REALISTIC" | "LOCAL_FIT_DIAGRAM";

export type VtonPrepareInput = {
  mannequin: MannequinParams;
  garment_processed_url?: string | null;
  garment_original_url?: string | null;
  person_image_url?: string | null;
  user_uploaded_person_image_url?: string | null;
  fit_zones: FitZone[];
};

export type VtonPayload = {
  mannequin: MannequinParams;
  garment_processed_url?: string | null;
  garment_original_url?: string | null;
  person_image_url?: string | null;
  user_uploaded_person_image_url?: string | null;
  fit_zones: FitZone[];
  render_mode: string;
  recommended_view_count: number;
  notes: string[];
  api_ready_payload: Record<string, unknown>;
};

export type VtonMockInput = {
  payload: VtonPayload;
};

export type VtonMockResult = {
  result_url: string;
  result_path: string;
  render_method: VtonRenderMethod;
  message: string;
};

export type VtonRunInput = {
  payload: VtonPayload;
  mode: VtonRunMode;
};

export type VtonRunResult = {
  result_url?: string | null;
  result_path?: string | null;
  provider: string;
  mode_requested: VtonRunMode;
  render_method: VtonRenderMethod;
  status?: string | null;
  used_fallback: boolean;
  success: boolean;
  message: string;
  raw_response?: Record<string, unknown> | null;
};

export type VtonTaskCreated = {
  task_id: string;
  state: VtonTaskState;
  poll_after_seconds: number;
  message: string;
};

export type VtonTaskStatusResponse = {
  task_id: string;
  state: VtonTaskState;
  result?: VtonRunResult | null;
  error?: string | null;
  poll_after_seconds: number;
};

export type PersonEphemeralUploadResult = {
  person_image_url: string;
  expires_in_seconds: number;
  message: string;
};

File: backend/app/models/vton.py
# -*- coding: utf-8 -*-
from pydantic import BaseModel
from typing import List, Optional, Dict, Any, Literal

from app.models.body import MannequinParams
from app.models.product import FitZone


VtonRunMode = Literal["mock", "external", "auto"]
VtonTaskState = Literal["queued", "running", "succeeded", "failed"]
VtonRenderMethod = Literal["NEURAL_REALISTIC", "LOCAL_FIT_DIAGRAM"]


class VtonPrepareInput(BaseModel):
    mannequin: MannequinParams
    garment_processed_url: Optional[str] = None
    garment_original_url: Optional[str] = None
    person_image_url: Optional[str] = None
    user_uploaded_person_image_url: Optional[str] = None
    fit_zones: List[FitZone] = []


class VtonPayload(BaseModel):
    mannequin: MannequinParams
    garment_processed_url: Optional[str]
    garment_original_url: Optional[str]
    person_image_url: Optional[str] = None
    user_uploaded_person_image_url: Optional[str] = None
    fit_zones: List[FitZone]
    render_mode: str
    recommended_view_count: int
    notes: List[str]
    api_ready_payload: Dict[str, Any]


class VtonMockInput(BaseModel):
    payload: VtonPayload


class VtonMockResult(BaseModel):
    result_url: str
    result_path: str
    render_method: VtonRenderMethod = "LOCAL_FIT_DIAGRAM"
    message: str


class VtonRunInput(BaseModel):
    payload: VtonPayload
    mode: VtonRunMode = "auto"


class VtonRunResult(BaseModel):
    result_url: Optional[str] = None
    result_path: Optional[str] = None
    provider: str
    mode_requested: VtonRunMode
    render_method: VtonRenderMethod
    status: Optional[str] = None
    used_fallback: bool
    success: bool
    message: str
    raw_response: Optional[Dict[str, Any]] = None


class VtonTaskCreated(BaseModel):
    task_id: str
    state: VtonTaskState
    poll_after_seconds: int = 2
    message: str


class VtonTaskStatusResponse(BaseModel):
    task_id: str
    state: VtonTaskState
    result: Optional[VtonRunResult] = None
    error: Optional[str] = None
    poll_after_seconds: int = 2


class PersonEphemeralUploadResult(BaseModel):
    person_image_url: str
    expires_in_seconds: int
    enhanced: bool = True
    message: str
